state-commitments/optimized/history_commitment.go

```diff
@@ -0,0 +1,489 @@
+	if err != nil {
+		return nil, err
+	}
+	return append(expac, emptyHash), nil
+}
+
+func (h *HistoryCommitter) proof(index uint64, leaves []common.Hash, virtual, limit uint64) (tail []common.Hash, err error) {
+	m := uint64(len(leaves))
+	if m == 0 {
+		return nil, errors.New("empty leaves slice")
+	}
+	if limit == 0 {
+		limit = nextPowerOf2(virtual)
+	}
+	if limit == 1 {
+		// Can only reach this with index == 0
+		return
+	}
+	if index >= limit/2 {
+		if m > limit/2 {
+			if limit/2 < m {
+				return h.proof(index-limit/2, leaves[limit/2:], virtual-limit/2, limit/2)
+			} else {
+				return nil, errors.New("invalid limit for given leaves")
+			}
+		}
+		if len(h.lastLeafFillers) > 0 {
+			return h.proof(index-limit/2, []common.Hash{h.lastLeafFillers[0]}, virtual-limit/2, limit/2)
+		} else {
+			return nil, errors.New("lastLeafFillers is empty")
+		}
+	}
+	if m > limit/2 {
+		tail, err = h.proof(index, leaves[:limit/2], limit/2, limit/2)
+		if err != nil {
+			return nil, err
+		}
+		right, err2 := h.subtreeExpansion(leaves[limit/2:], virtual-limit/2, limit/2, true)
+		if err2 != nil {
+			return nil, err2
+		}
+		for i := len(right) - 1; i >= 0; i-- {
+			tail = append(tail, right[i])
+		}
+		return tail, nil
+	}
+	if virtual > limit/2 {
+		tail, err = h.proof(index, leaves, limit/2, limit/2)
+		if err != nil {
+			return nil, err
+		}
+		if len(h.lastLeafFillers) > 0 {
+			right, err := h.subtreeExpansion([]common.Hash{h.lastLeafFillers[0]}, virtual-limit/2, limit/2, true)
+			if err != nil {
+				return nil, err
+			}
+			for i := len(right) - 1; i >= 0; i-- {
+				tail = append(tail, right[i])
+			}
+		} else {
+			return nil, errors.New("lastLeafFillers is empty")
+		}
+		return tail, nil
+	}
+	return h.proof(index, leaves, virtual, limit/2)
+}
+
+func (h *HistoryCommitter) prefixAndProof(index uint64, leaves []common.Hash, virtual uint64) (prefix []common.Hash, tail []common.Hash, err error) {
+	m := uint64(len(leaves))
+	if m == 0 {
+		return nil, nil, errors.New("nil leaves")
+	}
+	if virtual == 0 {
+		return nil, nil, errors.New("virtual size cannot be zero")
+	}
+	if m > virtual {
+		return nil, nil, fmt.Errorf("num leaves %d should be <= virtual %d", m, virtual)
+	}
+	if index+1 > virtual {
+		return nil, nil, fmt.Errorf("index %d + 1 should be <= virtual %d", index, virtual)
+	}
+
+	// Check for potential overflow before doing math.Log2
+	if virtual == 0 {
+		return nil, nil, errors.New("virtual size cannot be zero")
+	}
+	logVirtual := int(math.Log2(float64(virtual)) + 1)
+
+	// Ensure m > 0 before accessing leaves[m-1]
+	if m > 0 {
+		h.lastLeafFillers, err = h.precomputeRepeatedHashes(&leaves[m-1], logVirtual)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		return nil, nil, errors.New("leaves slice is empty")
+	}
+
+	if index+1 > m {
+		prefix, err = h.subtreeExpansion(leaves, index+1, 0, false)
+	} else {
+		// Ensure index+1 <= m before slicing
+		if index+1 <= m {
```

Comment:
This check is useless, we are in the `else` branch of an explicit `index + 1 > m` check